feat(files_sharing): Add toggle to control display area for federated shares

nfebe opened the pull request:
Add command to control display area for federated shares following : https://github.com/nextcloud/server/pull/52423

Usage: `occ config:app:set --value false --type boolean files_sharing show_federated_shares_as_internal`

### Screencast 

https://github.com/user-attachments/assets/b98c3f8d-31fd-45b5-aaaa-f55ebfb3611b

nfebe committed:
feat(files_sharing): Add toggle for federated share display area

Signed-off-by: nfebe <fenn25.fn@gmail.com>

nfebe commented on apps/files_sharing/lib/AppInfo/Application.php:
```diff
@@ -111,6 +114,9 @@ function () use ($c) {
 	public function boot(IBootContext $context): void {
 		$context->injectFn([$this, 'registerMountProviders']);
 		$context->injectFn([$this, 'registerEventsScripts']);
+		$context->injectFn(function (IInitialState $initialState, IConfig $config) {
```
I am unable to decide if this is the best place to provide this config.

kesselb replied:
As late as possible 😉

It should be possible to provide the initial state in apps/files_sharing/lib/Listener/LoadAdditionalListener.php.

Maybe even in apps/files_sharing/lib/Listener/LoadSidebarListener.php (if your change is only relevant for the sidebar).

kesselb commented on apps/files_sharing/lib/AppInfo/Application.php:
```diff
@@ -111,6 +114,9 @@ function () use ($c) {
 		$context->injectFn([$this, 'registerMountProviders']);
 		$context->injectFn([$this, 'registerEventsScripts']);
+		$context->injectFn(function (IInitialState $initialState, IConfig $config) {
+			$initialState->provideInitialState('showFederatedSharesAsInternal', $config->getSystemValue('sharing.federation.show_remote_shares_as_internal', false));
```
Please add `sharing.federation.show_remote_shares_as_internal`  to config.sample.php

nfebe replied:
Used app level config instead and added command to control this, so no need for that anymore. Thanks!

kesselb commented:
Hi,

With sharing.federation.show_remote_shares_as_internal = true, is it possible to add such a share by entering the user's name in the internal shares section? I currently don't have a GS setup to validate this myself.

Given that federation and GS are quite complex topics with various configuration options that can influence behavior, I think we should do another round of planning to adjust the new sharing sidebar to meet expectations. I'm concerned that a simple flag to move shares from one list to another might not be sufficient.

AndyScherzinger commented:
> With sharing.federation.show_remote_shares_as_internal = true, is it possible to add such a share by entering the user's name in the internal shares section? I currently don't have a GS setup to validate this myself.

Yes, that is the goal (or how it has been before the latest change). For a GS setup with this config flag (and it ultimately will do more than just this PR). In the end federated shared (in a GS setup/scenario) shall behave (from a user perspective) like local shares (while this will have its limitations or in other word won't be possible 100% given technical limitations).

So another (the second aspect to come later) is to hide the (remote) info on the username on the share-item in the sidebar because "should behave local".

AndyScherzinger reviewed (approved):
🦈

kesselb commented on apps/files_sharing/src/services/ConfigService.ts:
```diff
@@ -306,4 +307,14 @@ export default class Config {
+	 * @return {boolean}
+	 */
+	get showFederatedSharesAsInternal(): boolean {
+		const showFederatedSharesAsInternal = loadState('files_sharing', 'showFederatedSharesAsInternal', false)
+
+		return !!showFederatedSharesAsInternal
```
!! to map undefined to false? 
Why would it return undefined if there's a default with false?

kesselb commented:
> Iirc we treat federation different with global scale anyways so we could just check if global scale is enabled and if it is then we just treat any remote user as internal at least for the sharing sidebar maybe this could be an option so we don't need any UI option.[^1]

Currently, the administrator has to set the configuration value to revert to the old behavior.

I assume Ferdinand was referring to the following code. With GS enabled, the sharees API will query the lookup server regardless of the given share type, and therefore also return users from other nodes in the GS compound.

While I don't fully understand how onlyInternalFederation (gs.federation true/false) influences this, we could use isGlobalScaleEnabled as the default value to revert to the old behavior for affected customers without requiring any additional actions on their end. However, we should keep the option to override it for those who might need it.

https://github.com/nextcloud/server/blob/9a16e4fd1407f0490136f6cfeec184be4b219dce/apps/files_sharing/lib/Controller/ShareesAPIController.php#L179-L180

Example:

```php
$this->initialState->provideInitialState(
	'showFederatedSharesAsInternal',
	$this->appConfig->getAppValueBool('show_federated_shares_as_internal', $this->globalScaleConfig->isGlobalScaleEnabled())
);
```

[^1]: https://github.com/nextcloud/server/pull/52423#issuecomment-2834542402

ArtificialOwl committed:
refactor: use config lexicon

Signed-off-by: Maxence Lange <maxence@artificial-owl.com>

ArtificialOwl commented:
I have added Lexicon so we keep the description from Fon on the config value, also we enforce its type.
Still need to remove the occ command tho!

nfebe committed:
fix(files_sharing): Move `ConfigLexicon` to Config dir, update sharing input

- Fix autoloading for new `ConfigLexicon`
- Ensure that sharing input in sharing tab respect `show-federated-shares-as-internal`:
This is important, because when federated shares are shown as internal the users should add them from the internal shares section

Signed-off-by: nfebe <fenn25.fn@gmail.com>

nfebe commented:
> With sharing.federation.show_remote_shares_as_internal = true, is it possible to add such a share by entering the user's name in the internal shares section? I currently don't have a GS setup to validate this myself.

@kesselb yes, that is now respected!

nfebe commented:
/compile

maximelehericy reviewed (changes requested):
@nfebe @ArtificialOwl 
I am wondering about a case where some federated shares would be trusted and others not.
Typically, for MEJS, it is expected that shares from users from the lookup server end up in the internal sharing section.
However, if shares from users **outside ** of the lookup server appear in the internal sharing section, I expect users to be puzzled (see image below).

![image](https://github.com/user-attachments/assets/75b2b09d-9ca2-414f-80ff-476c1e045dc7)

I am not fully sure of the logic, your opinion would probably help here, but from an end-user perspective:

- seeing cloudIDs e.g. `alice@nc1.local.mlh.ovh` in the internal shares will looks weird
- seeing regular display names e.g. `bob` in external shares can have some sense in certain conditions.

With a logic table:

|Server of recipient is trusted ?|Recipient's displayName can be fetched from lookup server ?|Location of the share in the UI|Comment|
| --- | --- | --- | --- |
|0|0|External|Share to untrusted servers should always appear in external shares|
|0|1|External|Share to untrusted servers should always appear in external shares|
|1|0| depending on setting ?|I am torn, at first glance, I would say external because it looks weird to have cloudIDs in internal shares, but if for some reason the lookup server is gone, that might be useful|
|1|1|depending on setting| see [1]|

[1] Even if there is a displayName and the server is trusted, I expect certain organizations dealing with sensitive data would could use federation in the future to want to explicitely mark some shares as external to the instance. That is why I think it should depend on the setting.

@nfebe @ArtificialOwl WDYT ?

nfebe commented:
> @nfebe @ArtificialOwl WDYT ?

These sound like valid enhancements. However, the current implementation already provides a level of control that addresses the issues which originally led us to pause the backport of share splitting.

  - If a user or organization is already using the split sidebar and prefers that all types of federated shares (trusted and untrusted) appear under internal shares, they can retain that behavior.
  - If they prefer it the other way round, they can use the available setting to move those shares to the external section.

The other enhancements can and should be implemented after the release on the 7th.

Without the split and toggle, the situation is much worse.

maximelehericy reviewed (dismissed):
according to Fon's last comment, approving this setting that allows to revert a behavior introduced by a previous PR.

However, IMO further enhancements should be implemented to avoid confusing users having cloudIDs appearing in Internal Shares section or the other way around, displayNames appearing in External Shares section.

nfebe commented:
> so I fully agree with the first PR you made to force all federated shares to the External section.
> However, the new setting should not revert all of this, but just the shares which go to trusted servers.
> 
> In that case, we would need another name like show_federated_shares_to_trusted_servers_as_internal.
> 
> And if so, a condition on whether or not the recipient's server belongs to trusted servers.


Yes, but that is an addition, how is that blocking on this?

maximelehericy reviewed (approved):
This change looks fine.
However, in a federated setup of trusted instance and/or a global scale setup, this setting will not fulfill all the requirements.
Another setting will be needed so that all federated shares go to external shares (default behavior), except those targeting an instance among the trusted ones.

nfebe commented:
/compile

nfebe commented:
/compile

nfebe commented:
/compile

nfebe commented:
/compile

ArtificialOwl commented:
II would move `ConfigLexicon.php` to `lib/` to keep consistency with other apps (as `/lib/AppInfo/` was not accepted ;p)

nfebe commented:
/compile

nextcloud-command committed:
chore(assets): Recompile assets

Signed-off-by: nextcloud-command <nextcloud-command@users.noreply.github.com>

maximelehericy commented:
@nfebe is the setting documented elsewhere than here ?

nfebe commented:
> @nfebe is the setting documented elsewhere than here ?
https://github.com/nextcloud/documentation/pull/13316

skjnldsv commented:
Sorry, but that sounds a bit extreme to add a custom config for this, no?
